File: typescript/day01.ts
import { assert } from 'node:console';
import { slurpLines } from './input.js';
import * as R from 'ramda';

function parseElves(lines: string[]): number[][] {
    const elves: number[][] = [[]];
    for (const line of lines) {
        if (line == '') {
            elves.push([]);
        } else {
            elves[elves.length - 1].push(parseInt(line));
        }
    }
    return elves;
}

function top(n: number, xs: number[]): number[] {
    return [...xs].sort((a, b) => a - b).slice(-n);
}

async function partOneSolution(path: string): Promise<number> {
    const lines = await slurpLines(path);
    const elves = parseElves(lines);
    return Math.max(...elves.map(R.sum));
}

async function partTwoSolution(path: string): Promise<number> {
    const lines = await slurpLines(path);
    const elves = parseElves(lines);
    return R.sum(top(3, elves.map(R.sum)));
}

async function main() {
    const examplePath = '../examples/day01.txt';
    const inputPath = '../input/day01.txt';

    assert(await partOneSolution(examplePath) === 24000);
    console.log(`part one: ${await partOneSolution(inputPath)}`);

    assert(await partTwoSolution(examplePath) === 45000);
    console.log(`part two: ${await partTwoSolution(inputPath)}`);
}

main();

File: clojure/src/advent2022/day01.clj
(ns advent2022.day01
  (:require [clojure.java.io :as io]))

(defn slurp-lines [f]
  (with-open [rdr (io/reader f)]
    (vec (line-seq rdr))))

(defn partition-when
  "Partitions a collection by elements matching a predicate.
   Matching elements are dropped from the result."
  ([pred coll] (partition-when pred coll nil))
  ([pred coll acc]
   (if (empty? coll)
     acc
     (let [[taken dropped] (split-with (complement pred) coll)]
       (partition-when pred (rest dropped) (cons taken acc))))))

(defn load-input [f]
  (let [parse-int #(Integer/parseInt %)]
    (->> (slurp-lines f)
         (partition-when empty?)
         (map #(map parse-int %)))))

(defn solve-part-one [input]
  (->> input
       (map #(reduce + %))
       (apply max)))

(defn solve-part-two [input]
  (->> input
       (map #(reduce + %))
       sort
       (take-last 3)
       (reduce +)))

(comment
  (def example-input (load-input "../examples/day01.txt"))
  (def real-input (load-input "../input/day01.txt"))
  
  (solve-part-one example-input)
  (solve-part-one real-input)
  
  (solve-part-two example-input)
  (solve-part-two real-input)
  :rcf)
